workflow/controller/operator.go

```diff
@@ -490,6 +492,23 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 	}
 }
 
+func (woc *wfOperationCtx) releaseLocksForPendingShuttingdownWfs(ctx context.Context) bool {
+	if woc.GetShutdownStrategy().Enabled() && woc.wf.Status.Phase == wfv1.WorkflowPending {
+		if woc.controller.syncManager.ReleaseAll(woc.execWf) {
+			woc.log.WithFields(log.Fields{"key": woc.execWf.Name}).Info("Released all locks since this pending workflow is being shutdown")
+			// 1. If a workflow is being stopped, all exit handlers should still run.
+			// 2. If a workflow is being terminated, workflow should be marked as succeeded.
+			if woc.execWf.Spec.HasExitHook() && woc.GetShutdownStrategy().ShouldExecute(true) {
+				woc.markWorkflowRunning(ctx)
+				return true
+			}
+			woc.markWorkflowSuccess(ctx)
+			return true
+		}
+	}
```

Comment:
Trying to understand something: 
if we stop a pending workflow that wants to acquire a mutex/semaphore, are we allowing it to move to Running without acquiring the lock?

Reply:
Because if so, the user's exit handler might also do things that need exclusive access and we would be violating the mutex/semaphore

Reply:
I think it is incorrect behavior for stop. We can't assume the intent of exit handler/hook logic. It could be considered part of a critical code section by a user. The guarantee we must always make with workflow-level semaphores/mutexes is that at most N workflows will run at a given time. Argo stop should not change that.

We do have to decide: `argo stop` of a Pending workflow should either:
1. allow a workflow to acquire the mutex/semaphore normally, and then essentially just run exit handlers as soon as it becomes Running
2. or behave identically to `argo terminate` and never execute any part of the workflow

Given that `argo terminate` already handles cancelling a workflow without running any logic, I think 2 may be more desirable?

Reply:
If the user wants their exit handlers to always run but have their main logic protected by the semaphore/mutex, then the user can do the following:
1. use a **_template_** level mutex/semaphore for their entrypoint template (not workflow level)
2. do not protect exit handlers/hooks with any mutex/semaphore
3. Run `argo stop` against their Running workflows instead of terminate

Reply:
Conclusion after discussion with Jesse:
1. `argo terminate` of a pending workflow waiting on a lock, immediately transitions it to `Succeeded` without waiting for lock aquisition
2. `argo stop` of a pending workflow waiting on a lock, will acquire the lock normally, and reconcile normally